File: Data Structures & Algorithms/count-good-nodes-in-binary-tree/submission-4.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def goodNodes(self, root: TreeNode) -> int:
        
        return self.dfs(root, -100)
    
    def dfs(self, root, currMax):
        isgood = 0
        if not root:
            return 0
        
        if root.val >= currMax:
            isgood = 1
        
        #update the max
        currMax = max(currMax, root.val)
        
        return isgood + self.dfs(root.left, currMax) + self.dfs(root.right, currMax)
    #pass the new maximum along the tree

File: Data Structures & Algorithms/balanced-binary-tree/submission-4.cs
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left;
 *     public TreeNode right;
 *     public TreeNode(int val=0, TreeNode left=null, TreeNode right=null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

public class Solution {
    bool check;
    public bool IsBalanced(TreeNode root) {
        if (root is null) return true;
        check = true;
        dfs(root);
        return check;
    }

    public int dfs(TreeNode root){
        if (root is null) return 0;

        var left = dfs(root.left);
        var right = dfs(root.right);

        if(Math.Abs(left - right) > 1) check = false;

        return 1 + Math.Max(left, right);

        
    }
}
